### PR TITLE
[FLINK-5884] [table] Integrate time indicators for Table API & SQL

### PR DESCRIPTION
**Pull request by twalthr**
This PR introduces time indicators in the Table & SQL API. The PR splits the current row type into a logical and physical row type. Time indicators are part of every logical row and stay logical if they are just forwarded and not accessed for calculation. If a time attribute is passed to a function or arithmetic expression it will be materialized. A ProcessFunction can access the meta timestamp for materialization (code generation for this has to be implemented). Every logical plan includes time attributes now and looks more consistent:

```
val table = util.addTable[(Long, Int, String)]('long.rowtime, 'int, 'string)

val windowedTable = table
  .window(Session withGap 7.milli on 'long as 'w)
  .groupBy('w, 'string)
  .select('string, 'int.count)

val expected = unaryNode(
  "DataStreamAggregate",
  streamTableNode(0),
  term("groupBy", "string"),
  term(
    "window",
    SessionGroupWindow(
      WindowReference("w"),
      'long,
      7.milli)),
  term("select", "string", "COUNT(int) AS TMP_0")
)
```

Main changes of the current solution:

- Processing and rowtime time indicators can be named arbitrarily
- They can be defined as follows: stream.toTable(tEnv, 'long, 'int, 'string, 'proctime.proctime) or stream.toTable(tEnv, 'long.rowtime, 'int, 'string)
- In a streaming environment: if the "long" field is already defined in the record, it will not be read by the runtime. "long" always represents the timestamp of the row.
- In batch environment: "long" must be present in the record and will be read by the runtime.
- The table definition looks equivalent in both batch and streaming (better unification than current state)
- Internally row types are split up in a logical and a physical row type.
- The logical row type contains time indicators, the physical rowtime never contains time indicators (the pure "long" will never be in a record)
- After validation and query decorrelation, a special time indicator converter traverses the RelNodes and analyzes if the a time indicator is accessed or only forwarded.
- An access to a time indicator means that we need to materialize the rowtime using a ProcessFunction (not yet implemented). The timestamp (not an indicator anymore) becomes part of the physical row. E.g. long.cast(STRING) would require a materialization
- Forwarding of time indicators does not materialize the rowtime. It remains a logical attribute. E.g. .select('long)
- Windows are only valid if they work on time indicators.
- A new `RowSchema` unifies logical and physical rows and tries to make type handling throughout the classes easier and consistent (it is not used everywhere yet).

Missing so far:
- Java API tests for defining time attributes
- Tests for TableSources that define time attributes
- Multiwindow tests
- ProcessFunction for accessing the timestamp
- More tests in general

### CODE REVIEW

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/StreamTableSourceScan.scala [hunk 1]
```diff
@@ -37,7 +38,50 @@ class StreamTableSourceScan(
...
+      } else {
+        l.take(idx) ++ l.drop(idx + 1)
+      }
+    }
+
+    var fieldNames = TableEnvironment.getFieldNames(tableSource).toList
+    var fieldTypes = TableEnvironment.getFieldTypes(tableSource.getReturnType).toList
+
+    val rowtime = tableSource match {
+      case timeSource: DefinedTimeAttributes if timeSource.getRowtimeAttribute != null =>
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/datastream/StreamTableSourceScan.scala]: There are no tests for `TableSource`s that implement `DefinedTimeAttributes`

**Comment by fhueske** [comment]
I didn't look at `RelTimeIndicatorConverter` yet. 
Will do so tomorrow.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/BatchTableEnvironment.scala [hunk 2]
```diff
@@ -194,7 +194,30 @@ abstract class BatchTableEnvironment(
...
+    val (fieldNames, fieldIndexes) = getFieldInfo[T](
+      dataSet.getType,
+      fields,
+      ignoreTimeAttributes = true)
+
+    // validate and extract time attributes
+    val (rowtime, proctime) = validateAndExtractTimeAttributes(fieldNames, fieldIndexes, fields)
+
+    // don't allow proctime on batch
+    proctime match {
```
Comment (sunjincheng121) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/BatchTableEnvironment.scala]: In fact, In batch mode we only support `event-time`, And the the column that represents the `rowtime` is not an indicators but a real exist field. We only care about the data type of the field,So IMO. we can not using both `.proctime` and `.rowtime` in batch mode.

File: flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/StreamTableEnvironment.scala [hunk 3]
```diff
@@ -145,6 +105,7 @@ abstract class StreamTableEnvironment(
             "StreamTableEnvironment")
     }
   }
+
   /**
```
Comment (fhueske) [inline flink-libraries/flink-table/src/main/scala/org/apache/flink/table/api/StreamTableEnvironment.scala]: Time materialization needs to be applied to the output as well. This should solve the issue.

**Comment by fhueske** [comment]
Hi @twalthr, I addressed my comments, rebased to current master, and will merge this PR given that all tests pass.

Thanks, Fabian

**Pull request closed without merging**